### PR TITLE
Fixed #25991 -- New startegy to filter exclude() splitting

### PR DESCRIPTION
**Pull request by akaariai**
This commit tries to fix queries which filter against a multijoin multiple times in the same query, when at least one of the filter clauses is negated. Previously all negated multijoin filter clauses were pushed to separate subqueries, now they are pushed to same subquery if the filters target the same multijoin.

Example: qs.exclude(authors__age__gte=20, authors__age__lte=30): currently, this generates two different subqueries, after this pr, just one subquery is generated.

### CODE REVIEW

File: tests/queries/tests.py [hunk 1]
```diff
@@ -3669,3 +3671,66 @@ def test_ticket_24278(self):
         School.objects.create()
         qs = School.objects.filter(Q(pk__in=()) | Q())
         self.assertQuerysetEqual(qs, [])
+
+
+class ExcludeCombineTests(TestCase):
+    def setUp(self):
```
Comment (jarshwah) [inline tests/queries/tests.py]: Prefer setUpTestData to setUp

```
@classmethod
def setUpTestData(cls):
```

File: django/db/models/sql/query.py [hunk 2]
```diff
@@ -1202,6 +1191,89 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
     def add_filter(self, filter_clause):
         self.add_q(Q(**{filter_clause[0]: filter_clause[1]}))
 
+    def find_base_splits(self, q_object, found_splits, in_negated=False, path=None):
```
Comment (jarshwah) [inline django/db/models/sql/query.py]: This method is quite confusing to me. It's called recursively, but doesn't return anything. Is it just mutating the state passed in? Anyway to make this clearer?

**Comment by akaariai** [comment]
I think this one is functionally correct now. Unfortunately the code does need a lot of cleanup in some places.

I'll look into the review comments later on.

File: tests/queries/tests.py [hunk 3]
```diff
@@ -3686,3 +3686,75 @@ def test_invalid_values(self):
...
+        r1 = RelatedObject.objects.create(f=1)
+        r2 = RelatedObject.objects.create(f=2)
+        r3 = RelatedObject.objects.create(f=2)
+        Plaything.objects.create(others=r1, name='bar')
+        Plaything.objects.create(others=r2, name='bar')
+        Plaything.objects.create(others=r3, name='bar')
+        Plaything.objects.create(others=r3, name='foo')
+        # r3 doesn't match, so no count from r3 for f=2.
+        self.assertQuerysetEqual(
+            RelatedObject.objects.values('f').exclude(plaything__name='foo').annotate(plaything_cnt=Count('plaything')).order_by('plaything_cnt'),
```
Comment (charettes) [inline tests/queries/tests.py]: `.order_by('plaything_cnt', 'f')` should solve the test failure.

**Comment by akaariai** [comment]
buildbot, retest this please.

File: tests/queries/tests.py [hunk 4]
```diff
@@ -3800,3 +3800,77 @@ def test_ticket_23622(self):
             Ticket23605A.objects.filter(qx),
             [a2], lambda x: x
         )
+
+
+class ExcludeCombineTests(TestCase):
+    def setUp(self):
```
Comment (timgraham) [inline tests/queries/tests.py]: setUpTestData

File: tests/queries/tests.py [hunk 5]
```diff
@@ -3800,3 +3800,77 @@ def test_ticket_23622(self):
             Ticket23605A.objects.filter(qx),
             [a2], lambda x: x
         )
+
+
+class ExcludeCombineTests(TestCase):
```
Comment (timgraham) [inline tests/queries/tests.py]: Would it make sense to move this to a new `queries/test_xxx.py` file? I think we need to break up the file and maybe an incremental approach when adding new tests would work best?

Comment (akaariai) [inline tests/queries/tests.py]: I'd like to make a decision about this separately. I agree that the queries/tests.py file is hard to maintain, so we need to do something here.

**Comment by apollo13** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
Ready for a review now? Does this reference any existing tickets or should we create a new one?

**Comment by akaariai** [comment]
Yes, I think this is ready for review. I'll create a separate ticket for this. There are likely some existing tickets this solves, I'll try to find those and link in to the new ticket.

File: django/db/models/query_utils.py [hunk 6]
```diff
@@ -51,6 +51,7 @@ class Q(tree.Node):
     AND = 'AND'
     OR = 'OR'
     default = AND
+    subq_split_pos = False
```
Comment (timgraham) [inline django/db/models/query_utils.py]: Add comment about the meaning of this?

File: django/db/models/sql/query.py [hunk 7]
```diff
@@ -1099,8 +1092,8 @@ def try_transform(self, lhs, name, rest_of_lookups):
                 "permitted." %
                 (name, lhs.output_field.__class__.__name__))
 
-    def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
-                     can_reuse=None, connector=AND, allow_joins=True, split_subq=True):
+    def build_filter(self, filter_expr, current_negated=False,
```
Comment (timgraham) [inline django/db/models/sql/query.py]: I'll edit this docstring to remove the references to the removed parameers, but please check my edits.

**Comment by timgraham** [comment]
Cosmetic edits: http://dpaste.com/1NCAYSX
I wonder if we should add this to the release notes as I imagine some users might rely on the current "incorrect" behavior and maybe some documentation that describes the expected behavior.

**Comment by akaariai** [comment]
Yes, this one needs release notes as it can change query results for users.

**Comment by akaariai** [comment]
Rebased on master, release notes added and Tim's review modifications applied.

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
One test failure on Oracle looks like it could be an ordering issue:
```
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-oracle/database/oracle12/label/trusty-pr/python/python3.5/tests/queries/tests.py", line 3942, in test_exclude_values_annotate
    [{'f': 1, 'plaything_cnt': 1}, {'f': 2, 'plaything_cnt': 1}], lambda x: x
  File "/home/jenkins/workspace/pull-requests-oracle/database/oracle12/label/trusty-pr/python/python3.5/django/test/testcases.py", line 942, in assertQuerysetEqual
    return self.assertEqual(list(items), values, msg=msg)
AssertionError: Lists differ: [{'plaything_cnt': 1, 'f': 2}, {'plaything_cnt': 1, 'f': 1}] != [{'plaything_cnt': 1, 'f': 1}, {'plaything_cnt': 1, 'f': 2}]

First differing element 0:
{'plaything_cnt': 1, 'f': 2}
{'plaything_cnt': 1, 'f': 1}

- [{'f': 2, 'plaything_cnt': 1}, {'f': 1, 'plaything_cnt': 1}]
?        ^                             ^

+ [{'f': 1, 'plaything_cnt': 1}, {'f': 2, 'plaything_cnt': 1}]
?        ^                             ^
```

File: django/db/models/query_utils.py (hunk 6 above)
Comment (timgraham) [inline django/db/models/query_utils.py]: Any reply to this comment?

**Comment by timgraham** [comment]
Doc edits: http://dpaste.com/1ME7GAY 
In the topic guide, this restores some text removed when the note that's deleted here was added  (6770b7ecd).
Please double check that the advice in the release notes about constructing a compatible query is accurate (I tested it for one case and it seemed to work).

**Comment by akaariai** [comment]
I'll still check what exactly are the backwards incompatibility issues with .exclude().

I believe we might have a bit of problems for providing exactly the same behavior as Django 1.9 for complex Q-objects inside single .exclude() call.

**Comment by akaariai** [comment]
I still found some cases that do not work the way they should. So, this needs a bit more work...

File: django/db/models/sql/query.py [hunk 8]
```diff
@@ -1099,24 +1092,16 @@ def try_transform(self, lhs, name, rest_of_lookups):
...
         The 'current_negated' is used to determine if the current filter is
         negated or not and this will be used to determine if IS NULL filtering
         is needed.
 
-        The difference between current_netageted and branch_negated is that
-        branch_negated is set on first negation, but current_negated is
-        flipped for each negation.
-
         Note that add_filter will not do any negating itself, that is done
         upper in the code by add_q().
```
Comment (timgraham) [inline django/db/models/sql/query.py]: This was noted in another PR that I closed to avoid creating a merge conflict: "upper" -> "earlier"

**Comment by akaariai** [comment]
We need further large changes to allow exact similar queries than what we had in 1.9 for the double-exclude case.

The choices are to include this as is with a notion that for some queries that worked against the documented behavior in 1.9 *can not be achieved in 1.10*, or continue working on this branch without merging. I think I could make an additional branch for the further work.

**Comment by timgraham** [comment]
If you think merging this now will prevent the work from going stale and are optimistic about doing the further work for that case by the alpha (May 16), I'm okay with it.

**Comment by akaariai** [comment]
I'm trying to come up with a viable plan for fixing the issues with the current way of dealing with multivalued relations. This isn't looking too good, we have a bunch of really hard issues to solve. I'll leave this PR open for a while.

**Comment by akaariai** [comment]
I'm still working on a good solution to the backwards compat case.

What I have come up with is that .exclude() (which are similar to ~Q() queries) aren't that simple to implement for multivalued references. The problem is that .filter(some_cond) will return objects for which the some_cond matches for at least a single row in the related objects set. The .exclude(some_cond) removes rows for which no related object matches the some_cond. The end result is that:
  1) If you do Author.objects.exclude(friends__age__lte=30).annotate(Sum('friends__age')) you will *not* get the sum of ages for friends that are over 30 years old. The reason is that if an author has a single friend less than 30 years old that author is removed from the results, and thus the sum isn't calculated at all for the author.
  2) Author.objects.exclude(friends__age__lte=30, friends__height__lte=180) is actually the same query as Author.objects.exclude(friends__age__lte=30).exclude(friends__height__lte=180). The reason is that the age exclude either retains all related rows, or removes the author from the results. So, the second exclude() doesn't need to work on the same set of related rows, because the set is always unfiltered for any author still in the query.
  3) I am not yet sure how queries like .filter(Q(friends__age__lte=30) | ~Q(friends__height__lte=40)) should actually work.

**Comment by akaariai** [comment]
I'm having big problems with this PR. The main problem right now is a query like:
    Author.objects.exclude(Q(friends1__age__lte=30) | Q(friends2__age__height=40)).exclude(Q(friends1__height__lte=20) | Q(friends2__age__height=30))

Here friends1 refers to one "join" of friends, friends2 refers to another join of friends. For negated filters we need to push the conditions to subquery. But how to push the same join to two different subqueries?

I think the answer is to raise an exception on queries we can't handle correctly... But even that turns out to be hard to do correctly.

**Comment by akaariai** [comment]
See https://github.com/django/django/pull/6005 for a way to explicitly handle the join generation.

**Comment by akaariai** [comment]
I've been trying to come up with strategies that allow for correct, efficient and implementable ways to convert complex negated queries to SQL. I believe no such way actually exists, and for that reason it might be better to not even try to support complex negated queries.

**Comment by akaariai** [comment]
This issue is too complex to fix with the time I have available. I'll get back to this if I have more time to work on Django.

**Comment by charettes** [comment]
I'll try to revisit this approach in the next weeks.

Looking back at it the main blocker for pushing it forward was a lack of relationship aliasing which prevented some queries from being expressed with the proposed exclusion logic.

Now that we have an equivalent through `FilteredRelation` (unfortunate name as it would have been great to be able to create relationship aliases without including a filter and kind it's kind of an abuse of `annotate` which would have been better suited for `QuerySet.alias`) it should be possible to error out clearly in ambiguous case and suggest `FilteredRelation` be used.

**Pull request closed without merging**